php/class-utils.php

```diff
@@ -578,4 +578,106 @@ public static function svg_encoded( $width = '600px', $height = '400px', $color
 
 		return 'data:image/svg+xml;base64,' . base64_encode( $svg );
 	}
+
+	/**
+	 * Log a debug message.
+	 *
+	 * @param string      $message The message to log.
+	 * @param string|null $key     The key to log the message under.
+	 */
+	public static function log( $message, $key = null ) {
+		if ( get_plugin_instance()->get_component( 'report' )->enabled() ) {
+			$messages = get_option( Sync::META_KEYS['debug'], array() );
+			if ( $key ) {
+				$hash                      = md5( $message );
+				$messages[ $key ][ $hash ] = $message;
+			} else {
+				$messages[] = $message;
+			}
+			update_option( Sync::META_KEYS['debug'], $messages, false );
+		}
+	}
+
+	/**
+	 * Get the debug messages.
+	 *
+	 * @return array
+	 */
+	public static function get_debug_messages() {
+		return get_option( Sync::META_KEYS['debug'], array( __( 'Debug log is empty', 'cloudinary' ) ) );
+	}
+
+	/**
+	 * Check if the tag attributes contain possible third party manipulated data, and return found data.
+	 *
+	 * @param array $attributes The tag attributes.
+	 *
+	 * @return string|false
+	 */
+	public static function maybe_get_third_party_changes( $attributes ) {
+		static $filtered_keys, $filtered_classes;
+		$lazy_keys    = array(
+			'src',
+			'lazyload',
+			'lazy',
+			'loading',
+		);
+		$lazy_classes = array(
+			'lazyload',
+			'lazy',
+			'loading',
+		);
+		if ( ! $filtered_keys ) {
+			/**
+			 * Filter the keywords in data-* attributes on tags to be ignored from lazy-loading.
+			 *
+			 * @hook   cloudinary_ignored_data_keywords
+			 * @since  3.0.7
+			 *
+			 * @param $lazy_keys {array} The built-in ignore data-* keywords.
+			 *
+			 * @return {array}
+			 */
+			$filtered_keys = apply_filters( 'cloudinary_ignored_data_keywords', $lazy_keys );
+
+			/**
+			 * Filter the keywords in classes on tags to be ignored from lazy-loading.
+			 *
+			 * @hook   cloudinary_ignored_class_keywords
```

Comment:
<!-- general note -->
**📝 note:** @DavidCramer I'll update these on the `develop` branch.